File: backend/src/modules/transportation/vehicles/vehicles.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../../prisma/prisma.service';
import { fail, paginationHeader, success, successList, successWrite } from '../../../common/response/base-response';
import { NotificationsService } from '../notifications/notifications.service';

@Injectable()
export class VehiclesService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly notifications: NotificationsService,
  ) {}

  /** GET getAllVehicleType — vehicle-type lookup ({ Id, Type, Active, TransportationVehicles: [] }). */
  async getAllVehicleType() {
    const rows = await this.prisma.vehicleType.findMany({ orderBy: { id: 'desc' } });
    const data = rows.map((t) => ({
      Id: t.id,
      Type: t.type,
      Active: t.active,
      TransportationVehicles: [] as unknown[],
    }));
    return success(data);
  }

  /** POST AddVehicleType — create a vehicle type. */
  async addVehicleType(type: string, active?: boolean) {
    if (!type?.trim()) return fail('Err101', 'Type is required');
    const created = await this.prisma.vehicleType.create({
      data: { type, active: active ?? true },
    });
    return successWrite(created.id);
  }

  /** POST UpdateVehicleType — rename / toggle a vehicle type. */
  async updateVehicleType(id: number, type: string, active?: boolean) {
    if (!id) return fail('Err101', 'Id is required');
    if (!type?.trim()) return fail('Err101', 'Type is required');
    await this.prisma.vehicleType.update({
      where: { id },
      data: { type, active: active ?? true },
    });
    return successWrite(id);
  }

  /** GET getAllTransportationVehicle — paginated vehicles with joined type name. */
  async getAllVehicles(pageNo = 1, noOfItems = 20) {
    const where = { active: true };
    const [total, rows] = await this.prisma.$transaction([
      this.prisma.transportationVehicle.count({ where }),
      this.prisma.transportationVehicle.findMany({
        where,
        include: { vehicleType: true },
        orderBy: { id: 'desc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);
    const data = rows.map((v) => ({
      Id: v.id,
      VehicleTypeId: v.vehicleTypeId,
      Capacity: v.capacity,
      IsApproved: v.isApproved,
      ApprovedBy: v.approvedBy ?? '',
      Active: v.active,
      CreationDate: v.creationDate,
      VehicleTypeName: v.vehicleType?.type ?? '',
      TransportationLines: [] as unknown[],
    }));
    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** POST AddTransportationVehicle — create (IsApproved defaults false). */
  async addVehicle(vehicleTypeId: number, capacity: number, active: boolean | undefined, userId: number) {
    if (!vehicleTypeId) return fail('Err101', 'VehicleTypeId is required');
    const created = await this.prisma.transportationVehicle.create({
      data: {
        vehicleTypeId,
        capacity: capacity ?? 0,
        active: active ?? true,
        isApproved: false,
        creationBy: userId,
        modifiedBy: userId,
      },
    });
    await this.notifications.notifySuperAdmins(userId, {
      title: 'إضافة مركبة جديدة',
      description: 'بحاجة لاعتماد',
      entityType: 'vehicle',
      entityId: created.id,
    });
    return successWrite(created.id);
  }

  /** POST UpdateTransportationVehicle — update type/capacity/active. */
  async updateVehicle(
    id: number,
    vehicleTypeId: number,
    capacity: number,
    active: boolean | undefined,
    userId: number,
  ) {
    if (!id) return fail('Err101', 'Id is required');
    if (!vehicleTypeId) return fail('Err101', 'VehicleTypeId is required');
    await this.prisma.transportationVehicle.update({
      where: { id },
      data: {
        vehicleTypeId,
        capacity: capacity ?? 0,
        active: active ?? true,
        modifiedBy: userId,
      },
    });
    return successWrite(id);
  }

  /** POST DeleteTransportationVehicle — delete (Id header). */
  async removeVehicle(id: number) {
    if (!id) return fail('Err101', 'Id is required');
    await this.prisma.transportationVehicle.delete({ where: { id } });
    return successWrite(id);
  }

  /** POST ApproveTransportationVehicle — approve/unapprove (Id + Approve headers). */
  async approveVehicle(id: number, approve: boolean, userId: number) {
    if (!id) return fail('Err101', 'Id is required');
    await this.prisma.transportationVehicle.update({
      where: { id },
      data: { isApproved: approve, approvedBy: String(userId) },
    });
    return successWrite(id);
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Vehicles/VehiclesService.cs
using GarasTransport.Api.Common;
using GarasTransport.Api.Modules.Transportation.Notifications;
using Microsoft.EntityFrameworkCore;

namespace GarasTransport.Api.Modules.Transportation.Vehicles;

public class VehiclesService
{
    private readonly AppDbContext _db;
    private readonly NotificationsService _notifications;

    public VehiclesService(AppDbContext db, NotificationsService notifications)
    {
        _db = db;
        _notifications = notifications;
    }

    /// <summary>GET getAllVehicleType — vehicle-type lookup.</summary>
    public async Task<object> GetAllVehicleType()
    {
        var rows = await _db.VehicleTypes.OrderByDescending(t => t.Id).ToListAsync();
        var data = rows.Select(t => new
        {
            Id = t.Id,
            Type = t.Type,
            Active = t.Active,
            TransportationVehicles = Array.Empty<object>(),
        });
        return Envelope.Success(data);
    }

    /// <summary>POST AddVehicleType — create a vehicle type.</summary>
    public async Task<object> AddVehicleType(string? type, bool? active)
    {
        if (string.IsNullOrWhiteSpace(type)) return Envelope.Fail("Err101", "Type is required");
        var created = new VehicleType { Type = type, Active = active ?? true };
        _db.VehicleTypes.Add(created);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(created.Id);
    }

    /// <summary>POST UpdateVehicleType — rename / toggle a vehicle type.</summary>
    public async Task<object> UpdateVehicleType(int id, string? type, bool? active)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        if (string.IsNullOrWhiteSpace(type)) return Envelope.Fail("Err101", "Type is required");
        var vt = await _db.VehicleTypes.FirstOrDefaultAsync(t => t.Id == id);
        if (vt == null) return Envelope.Fail("Err142", "Vehicle type not found");
        vt.Type = type;
        vt.Active = active ?? true;
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>GET getAllTransportationVehicle — paginated vehicles with joined type name.</summary>
    public async Task<object> GetAllVehicles(int pageNo = 1, int noOfItems = 20)
    {
        var query = _db.TransportationVehicles.Where(v => v.Active);
        var total = await query.CountAsync();
        var rows = await query
            .Include(v => v.VehicleType)
            .OrderByDescending(v => v.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .ToListAsync();

        var data = rows.Select(v => new
        {
            Id = v.Id,
            VehicleTypeId = v.VehicleTypeId,
            Capacity = v.Capacity,
            IsApproved = v.IsApproved,
            ApprovedBy = v.ApprovedBy ?? string.Empty,
            Active = v.Active,
            CreationDate = v.CreationDate,
            VehicleTypeName = v.VehicleType?.Type ?? string.Empty,
            TransportationLines = Array.Empty<object>(),
        });
        return Envelope.SuccessList(data, Envelope.Pagination(pageNo, noOfItems, total));
    }

    /// <summary>POST AddTransportationVehicle — create (IsApproved defaults false).</summary>
    public async Task<object> AddVehicle(int vehicleTypeId, int capacity, bool? active, int userId)
    {
        if (vehicleTypeId == 0) return Envelope.Fail("Err101", "VehicleTypeId is required");
        var created = new TransportationVehicle
        {
            VehicleTypeId = vehicleTypeId,
            Capacity = capacity,
            Active = active ?? true,
            IsApproved = false,
            CreationBy = userId,
            ModifiedBy = userId,
        };
        _db.TransportationVehicles.Add(created);
        await _db.SaveChangesAsync();
        await _notifications.NotifySuperAdmins(userId, new NotifyOptions
        {
            Title = "إضافة مركبة جديدة",
            Description = "بحاجة لاعتماد",
            EntityType = "vehicle",
            EntityId = created.Id,
        });
        return Envelope.SuccessWrite(created.Id);
    }

    /// <summary>POST UpdateTransportationVehicle — update type/capacity/active.</summary>
    public async Task<object> UpdateVehicle(int id, int vehicleTypeId, int capacity, bool? active, int userId)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        if (vehicleTypeId == 0) return Envelope.Fail("Err101", "VehicleTypeId is required");
        var v = await _db.TransportationVehicles.FirstOrDefaultAsync(x => x.Id == id);
        if (v == null) return Envelope.Fail("Err142", "Vehicle not found");
        v.VehicleTypeId = vehicleTypeId;
        v.Capacity = capacity;
        v.Active = active ?? true;
        v.ModifiedBy = userId;
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>POST DeleteTransportationVehicle — delete (Id header).</summary>
    public async Task<object> RemoveVehicle(int id)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var v = await _db.TransportationVehicles.FirstOrDefaultAsync(x => x.Id == id);
        if (v == null) return Envelope.Fail("Err142", "Vehicle not found");
        _db.TransportationVehicles.Remove(v);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>POST ApproveTransportationVehicle — approve/unapprove (Id + Approve headers).</summary>
    public async Task<object> ApproveVehicle(int id, bool approve, int userId)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var v = await _db.TransportationVehicles.FirstOrDefaultAsync(x => x.Id == id);
        if (v == null) return Envelope.Fail("Err142", "Vehicle not found");
        v.IsApproved = approve;
        v.ApprovedBy = userId.ToString();
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }
}
